### PR TITLE
Refactor Backoff to allows min below 1s

### PR DESCRIPTION
**Pull request by antoineco**
Also refactor tests to provide more control over the test logic

### CODE REVIEW

File: pkg/adapter/common/backoff_test.go [hunk 1]
```diff
@@ -109,15 +108,16 @@ func TestDuration(t *testing.T) {
...
 		fn         func(context.Context) (bool, error)
 		waitReturn bool
 		err        error
 	}{
-		"force stop with no errs": {
+		"force stop with no err": {
+			fn: func(ctx context.Context) (bool, error) {
+				// never terminates by itself
+				return true, nil
+			},
```
Comment (antoineco) [inline pkg/adapter/common/backoff_test.go:120]: I encountered a race when `fn` was `nil` and the `doneCh` wasn't closed fast enough.

File: pkg/adapter/common/backoff.go [hunk 2]
```diff
@@ -68,9 +69,9 @@ func NewBackoff(args ...time.Duration) *Backoff {
 	return backoff
 }
 
-// Duration can be used to get exponential backoff duration calculated for each new step
+// Duration returns the exponential backoff duration calculated for the current step.
 func (b *Backoff) Duration() time.Duration {
-	dur := time.Duration(float64(b.min)*math.Pow(b.factor, float64(b.step)) - float64(1*time.Second))
+	dur := time.Duration(float64(b.min)*math.Pow(b.factor, float64(b.step)) - float64(b.min))
```
Comment (antoineco) [inline pkg/adapter/common/backoff.go:74]: This change is what makes it work for min backoff values below 1s.

File: pkg/adapter/common/backoff_test.go [hunk 3]
```diff
@@ -127,47 +127,90 @@ func TestRun(t *testing.T) {
 			waitReturn: true,
 			err:        assert.AnError,
 		},
-		"fn resets backoff duration": {
-			fn: func(ctx context.Context) (bool, error) {
-				counter++
-				if counter > 3 {
-					if time.Since(start).Round(time.Second) != 3*time.Second {
```
Comment (antoineco) [inline pkg/adapter/common/backoff_test.go:134]: This time-based expectation was very optimistic, considering `start` is shared between all tests, and delaying one previous test could have  made this one fail.

I rewrote it using a channel to control the number of executions of the function from the main test logic.

**Review by antoineco — COMMENTED** [review]
Notes

File: pkg/adapter/common/backoff_test.go (hunk 3 above)
Comment (tzununbekov) [reply pkg/adapter/common/backoff_test.go:134]: nice, I remember I was having a hard time with this test and the time-based check was a sort of desperate decision

**Review by sebgoa — APPROVED** [review]
nothing is standing out after a quick check

File: pkg/adapter/common/backoff_test.go [hunk 4]
```diff
@@ -127,47 +128,89 @@ func TestRun(t *testing.T) {
...
+		// of the main test logic.
+		latch := make(chan struct{})
+
+		fn := func(ctx context.Context) (bool, error) {
+			latch <- struct{}{}
+
+			defer func() {
+				runCounter++
+				latch <- struct{}{}
+			}()
```
Comment (tzununbekov) [inline pkg/adapter/common/backoff_test.go:181]: I'm not sure that I completely understand why do we need that deferred function with the counter, but the thing with double locking latches is something that I want to use somewhere else :slightly_smiling_face:

**Review by tzununbekov — APPROVED** [review]

File: pkg/adapter/common/backoff_test.go (hunk 4 above)
Comment (antoineco) [reply pkg/adapter/common/backoff_test.go:181]: Nice catch, this is a leftover from an older version of this code! We only need to defer the "unlatch" here.

Comment (antoineco) [reply pkg/adapter/common/backoff_test.go]: Removed

**Pull request merged**